src/rules/prefer_const.rs

```diff
@@ -0,0 +1,854 @@
+    if var_decl.kind != VarDeclKind::Let {
+      return;
+    }
+
+    for decl in &var_decl.decls {
+      self.extract_decl_idents(&decl.name, decl.init.is_some(), false);
+    }
+  }
+
+  fn visit_assign_expr(
+    &mut self,
+    assign_expr: &AssignExpr,
+    _parent: &dyn Node,
+  ) {
+    assign_expr.visit_children_with(self);
+    match &assign_expr.left {
+      PatOrExpr::Pat(pat) => self.extract_assign_idents(&**pat),
+      PatOrExpr::Expr(_) => {}
+    };
+  }
+
+  fn visit_update_expr(
+    &mut self,
+    update_expr: &UpdateExpr,
+    _parent: &dyn Node,
+  ) {
+    match &*update_expr.arg {
+      Expr::Ident(ident) => {
+        self.check_declared_in_outer_scope(ident);
+        self.mark_reassigned(ident, false);
+      }
+      otherwise => otherwise.visit_children_with(self),
+    }
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::test_util::*;
+
+  // Some tests are derived from
+  // https://github.com/eslint/eslint/blob/v7.10.0/tests/lib/rules/prefer-const.js
+  // MIT Licensed.
+
+  #[test]
+  fn prefer_const_valid() {
+    assert_lint_ok_n::<PreferConst>(vec![
+      r#"var x = 0;"#,
+      r#"let x;"#,
+      r#"let x; { x = 0; } foo(x);"#,
+      r#"let x = 0; x = 1;"#,
+      r#"const x = 0;"#,
+      r#"for (let i = 0, end = 10; i < end; ++i) {}"#,
+      r#"for (let i in [1,2,3]) { i = 0; }"#,
+      r#"for (let x of [1,2,3]) { x = 0; }"#,
+      r#"(function() { var x = 0; })();"#,
+      r#"(function() { let x; })();"#,
+      r#"(function() { let x; { x = 0; } foo(x); })();"#,
+      r#"(function() { let x = 0; x = 1; })();"#,
+      r#"(function() { const x = 0; })();"#,
+      r#"(function() { for (let i = 0, end = 10; i < end; ++i) {} })();"#,
+      r#"(function() { for (let i in [1,2,3]) { i = 0; } })();"#,
+      r#"(function() { for (let x of [1,2,3]) { x = 0; } })();"#,
+      r#"(function(x = 0) { })();"#,
+      r#"let a; while (a = foo());"#,
+      r#"let a; do {} while (a = foo());"#,
+      r#"let a; for (; a = foo(); );"#,
+      r#"let a; for (;; ++a);"#,
+      r#"let a; for (const {b = ++a} in foo());"#,
+      r#"let a; for (const {b = ++a} of foo());"#,
+      r#"let a; for (const x of [1,2,3]) { if (a) {} a = foo(); }"#,
+      r#"let a; for (const x of [1,2,3]) { a = a || foo(); bar(a); }"#,
+      r#"let a; for (const x of [1,2,3]) { foo(++a); }"#,
+      r#"let a; function foo() { if (a) {} a = bar(); }"#,
+      r#"let a; function foo() { a = a || bar(); baz(a); }"#,
+      r#"let a; function foo() { bar(++a); }"#,
+      r#"
+    let id;
+    function foo() {
+        if (typeof id !== 'undefined') {
+            return;
+        }
+        id = setInterval(() => {}, 250);
+    }
+    foo();
+  "#,
+      r#"/*exported a*/ let a; function init() { a = foo(); }"#,
+      // TODO(magurotuna): this is ported from ESLint, but I have no idea why this is valid,
+      // so comment it out for now.
+      // r#"/*exported a*/ let a = 1"#,
```

Comment:
As I commented, I have no idea why this belongs to "valid" tests :confused: 
https://github.com/eslint/eslint/blob/v7.10.0/tests/lib/rules/prefer-const.js#L70

Reply:
@magurotuna this is probably some "magical" ESLint directive that makes it valid - we don't have such thing so you can safely remove this test :)

Reply:
oh, thanks, I didn't know that. I'll remove it.